# Importing Dataset and Libraries

In [1]:
import numpy as np
import pandas as pd

In [2]:
df = 'data2.xls'

In [3]:
td = pd.read_excel(df , sheet_name = 'transac_details')
td.head(2)

Credit_card  Product_ID P_CATEGORY CONDTION  Brand   Price   Selling_price  \
0       4595.0         534      SHOES      New  RKFCM  1484.74        1471.74   
1       6679.0         396    LUGGAGE      New  TNVON  1452.50        1442.50   

  Coupon_ID       Date      Time  ...    M_ID          Payment Method  \
0     BN710 2014-08-26  17:16:17  ...  YE5901            Prepaid card   
1     KF815 2014-01-26  15:40:52  ...  UW7354  Mobile carrier Billing   

  Transaction ID Return_ind Return date  new amount  Unnamed: 19 Unnamed: 20  \
0          30462          0  2014-09-02     1471.74          NaN         NaN   
1          37927          0  2014-02-02     1442.50          NaN         NaN   

   Discount Hour  
0      13.0   17  
1      10.0   15  

[2 rows x 23 columns]

In [4]:
ci = pd.read_excel(df , sheet_name = 'customer_info')
ci.head(2)

C_ID                Email   Name  Mobile_number Gender  Age        City   \
0  3768  eddie@yahoomail.com  EDDIE     9045258449      M   83   Louisville   
1  4852     rose@hotmail.com   ROSE     8834789103      F   87  Kansas City   

      State Address  
0  Kentucky  Masked  
1  Missouri  Masked

 ### Removing Blank Entries

In [5]:
td['Credit_card'].isnull().sum()  # Number of Blank Credit Cards Entries

5

In [6]:
td.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9999 entries, 0 to 9998
Data columns (total 23 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Credit_card     9994 non-null   float64       
 1   Product_ID      9999 non-null   int64         
 2   P_CATEGORY      9999 non-null   object        
 3   CONDTION        9999 non-null   object        
 4   Brand           9999 non-null   object        
 5   Price           9999 non-null   float64       
 6   Selling_price   9999 non-null   float64       
 7   Coupon_ID       9979 non-null   object        
 8   Date            9999 non-null   datetime64[ns]
 9   Time            9999 non-null   object        
 10  GTIN            9999 non-null   int64         
 11  MPN             9999 non-null   int64         
 12  Merchant_name   9999 non-null   object        
 13  M_ID            9999 non-null   object        
 14  Payment Method  9999 non-null   object        
 15  Tran

##### Since only 5 entries are missing out of 9999, we can delete those blank entries.

In [7]:
td = td.dropna(subset=['Credit_card'])

In [8]:
td.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9994 entries, 0 to 9998
Data columns (total 23 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Credit_card     9994 non-null   float64       
 1   Product_ID      9994 non-null   int64         
 2   P_CATEGORY      9994 non-null   object        
 3   CONDTION        9994 non-null   object        
 4   Brand           9994 non-null   object        
 5   Price           9994 non-null   float64       
 6   Selling_price   9994 non-null   float64       
 7   Coupon_ID       9974 non-null   object        
 8   Date            9994 non-null   datetime64[ns]
 9   Time            9994 non-null   object        
 10  GTIN            9994 non-null   int64         
 11  MPN             9994 non-null   int64         
 12  Merchant_name   9994 non-null   object        
 13  M_ID            9994 non-null   object        
 14  Payment Method  9994 non-null   object        
 15  Tran

# Tasks

## Customer Segmentation

In [16]:
def age_group(a,g):
    if g=='F':
        if a >= 18 and a <= 30:
            return 'Young Females'
        elif a>=31 and a<=45:
            return 'Mid-age Females'
        else:
            return 'Old Females'
    else:
        if a>=18 and a<=30:
            return 'Young Males'
        elif a>=31 and a<=45:
            return 'Mid-age Males'
        else:
            return 'Old Males'

In [19]:
ci['Age_Group']=ci.apply(lambda x:age_group(x['Age'],x['Gender']),axis=1)

In [20]:
ci.head()

C_ID                Email      Name  Mobile_number Gender  Age  \
0  3768  eddie@yahoomail.com     EDDIE     9045258449      M   83   
1  4852     rose@hotmail.com      ROSE     8834789103      F   87   
2  1174    amy@yahoomail.com       AMY     9557690013      F   31   
3  4807   clarence@gmail.com  CLARENCE     9394398429      M   37   
4  9131   johnny@hotmail.com    JOHNNY     9976623538      F   80   

         City        State Address        Age_Group  
0   Louisville    Kentucky  Masked        Old Males  
1  Kansas City    Missouri  Masked      Old Females  
2      Seattle  Washington  Masked  Mid-age Females  
3      Seattle  Washington  Masked    Mid-age Males  
4     Columbus        Ohio  Masked      Old Females

## Segmentwise Spend in Dollars

In [22]:
ci['Spend_Amt']=td.groupby('Credit_card')['Selling_price'].sum().reset_index()['Selling_price']

In [23]:
ci['Swipes']=td.groupby('Credit_card').count().reset_index()['Selling_price']

In [24]:
ci.head()

C_ID                Email      Name  Mobile_number Gender  Age  \
0  3768  eddie@yahoomail.com     EDDIE     9045258449      M   83   
1  4852     rose@hotmail.com      ROSE     8834789103      F   87   
2  1174    amy@yahoomail.com       AMY     9557690013      F   31   
3  4807   clarence@gmail.com  CLARENCE     9394398429      M   37   
4  9131   johnny@hotmail.com    JOHNNY     9976623538      F   80   

         City        State Address        Age_Group  Spend_Amt  Swipes  
0   Louisville    Kentucky  Masked        Old Males   56445.78      28  
1  Kansas City    Missouri  Masked      Old Females  154250.21      49  
2      Seattle  Washington  Masked  Mid-age Females  156280.67      49  
3      Seattle  Washington  Masked    Mid-age Males  190327.90     105  
4     Columbus        Ohio  Masked      Old Females  115323.46      49

In [27]:
ss=ci.groupby('Age_Group')['Spend_Amt'].sum().sort_values(ascending=False).reset_index()

In [28]:
ss

Age_Group   Spend_Amt
0      Old Females  9570645.85
1        Old Males  8409061.37
2    Young Females  2204630.27
3    Mid-age Males  2196427.54
4  Mid-age Females  2153508.46
5      Young Males  1726673.91

## Segmentwise number of swipes

In [29]:
ssw=ci.groupby('Age_Group')['Swipes'].sum().sort_values(ascending=False).reset_index()

In [30]:
ssw

Age_Group  Swipes
0      Old Females    3650
1        Old Males    3248
2    Mid-age Males     888
3  Mid-age Females     798
4    Young Females     761
5      Young Males     649

## Customer spend per product category

In [31]:
ps=td.groupby('P_CATEGORY')['Selling_price'].sum().sort_values(ascending=False).reset_index()
ps

P_CATEGORY  Selling_price
0              DECOR     3200896.29
1          COMPUTERS     3056561.16
2        ELECTRONICS     2684512.31
3    OFFICE SUPPLIES     2466129.19
4              SHOES     2295412.09
5   KITCHEN & DINING     2017038.89
6           CLOTHING     1972876.86
7            BEDDING     1661701.45
8              GAMES     1515700.09
9         APPLIANCES     1510925.54
10           LUGGAGE     1422861.12
11     BABY CLOTHING     1342503.30
12         BABY TOYS     1113829.11

##  Customer Spend per State

In [32]:
sst=ci.groupby('State')['Spend_Amt'].sum().sort_values(ascending=False).reset_index()
sst

State   Spend_Amt
0     California  7732818.88
1          Texas  4805470.75
2       Illinois  2459820.30
3     Washington  2093212.05
4       Kentucky  1936131.53
5         Nevada  1927880.03
6        Arizona  1823014.48
7  Massachusetts  1397459.88
8           Ohio  1121616.11
9       Missouri   963523.39

## Customer spend per Payment Method

In [34]:
sp=td.groupby('Payment Method')['Selling_price'].sum().sort_values(ascending=False).reset_index()
sp

Payment Method  Selling_price
0  Mobile carrier Billing    10851668.72
1             Credit card     9520888.13
2            Prepaid card     4497590.79
3              Debit card      831404.75
4           Paypal wallet      223567.96
5               Gift card      182324.68
6          Deirect debits      153502.37

## Five highest spending product category

In [36]:
ps=td.groupby('P_CATEGORY')['Selling_price'].sum().sort_values(ascending=False).reset_index()
ps.head()

P_CATEGORY  Selling_price
0            DECOR     3200896.29
1        COMPUTERS     3056561.16
2      ELECTRONICS     2684512.31
3  OFFICE SUPPLIES     2466129.19
4            SHOES     2295412.09

## Five highest spending States

In [37]:
sst=ci.groupby('State')['Spend_Amt'].sum().sort_values(ascending=False).reset_index()
sst.head()

State   Spend_Amt
0  California  7732818.88
1       Texas  4805470.75
2    Illinois  2459820.30
3  Washington  2093212.05
4    Kentucky  1936131.53

## Five highest spending Payment method

In [38]:
sp=td.groupby('Payment Method')['Selling_price'].sum().sort_values(ascending=False).reset_index()
sp.head()

Payment Method  Selling_price
0  Mobile carrier Billing    10851668.72
1             Credit card     9520888.13
2            Prepaid card     4497590.79
3              Debit card      831404.75
4           Paypal wallet      223567.96

## Returns made by each state

In [61]:
df=td.groupby('Credit_card')['Return_ind'].sum().reset_index()
df

Credit_card  Return_ind
0         1093.0           4
1         1174.0          10
2         1177.0           8
3         1256.0          16
4         1418.0           6
..           ...         ...
193       9756.0           4
194       9856.0           8
195       9896.0           5
196       9940.0          19
197       9950.0          17

[198 rows x 2 columns]

In [62]:
df1=df
df1['State']=ci['State']


In [63]:
dff1=df1.groupby('State')['Return_ind'].sum().sort_values(ascending=False).reset_index()
dff1

State  Return_ind
0     California         433
1          Texas         280
2       Illinois         145
3     Washington         125
4         Nevada         122
5       Kentucky         116
6        Arizona         112
7  Massachusetts          79
8           Ohio          69
9       Missouri          51

## Returns made by each age group

In [64]:
df2=df
df2['Age_Group']=ci['Age_Group']

In [65]:
dff2=df2.groupby('Age_Group')['Return_ind'].sum().sort_values(ascending=False).reset_index()
dff2

Age_Group  Return_ind
0      Old Females         573
1        Old Males         475
2    Mid-age Males         130
3  Mid-age Females         125
4    Young Females         125
5      Young Males         104

## Returns made for each Product Condition

In [66]:
cd=td.groupby('CONDTION')['Return_ind'].sum().sort_values(ascending=False).reset_index()
cd

CONDTION  Return_ind
0          New         712
1         Used         433
2  Refurbished         387

## Returns made for each Product Category

In [67]:
pd=td.groupby('P_CATEGORY')['Return_ind'].sum().sort_values(ascending=False).reset_index()
pd

P_CATEGORY  Return_ind
0              DECOR         182
1          COMPUTERS         171
2        ELECTRONICS         159
3              SHOES         141
4    OFFICE SUPPLIES         122
5   KITCHEN & DINING         120
6           CLOTHING         119
7              GAMES         113
8            LUGGAGE          95
9         APPLIANCES          92
10           BEDDING          83
11     BABY CLOTHING          73
12         BABY TOYS          62

## Returns made per Discount Value

In [68]:
dt=td.groupby('Discount')['Return_ind'].sum().sort_values(ascending=False).reset_index()
dt

Discount  Return_ind
0       26.0          59
1       23.0          49
2       37.0          47
3       29.0          47
4       44.0          46
5       49.0          45
6       18.0          43
7       32.0          43
8       31.0          42
9       19.0          40
10      46.0          40
11      15.0          38
12      38.0          38
13      20.0          38
14      16.0          37
15      21.0          37
16      22.0          37
17      13.0          37
18      43.0          37
19      35.0          36
20      48.0          36
21      41.0          36
22      12.0          36
23      34.0          35
24      47.0          35
25      36.0          34
26      17.0          33
27      28.0          32
28      50.0          32
29      30.0          30
30      24.0          30
31      11.0          30
32      42.0          29
33      14.0          29
34       0.0          29
35      39.0          28
36      10.0          28
37      40.0          28
38      45.0          27
39      25.0          27
40      33.0          26
41      27.0          25
42      90.0           3
43      50.0           3
44      49.0           2
45      47.0           2
46      37.0           2
47      35.0           1
48      25.0           1
49      22.0           1
50      36.0           1
51      49.0           1
52      22.0           1
53      43.0           1
54      16.0           1
55      47.0           1
56      27.0           0
57      37.0           0

## Customers' order timings

In [69]:
h=td.groupby('Hour')['Credit_card'].count().sort_values(ascending=False).reset_index()
h

Hour  Credit_card
0     11          553
1     17          509
2     12          488
3      3          481
4      7          476
5     21          462
6      8          462
7     22          447
8     18          446
9     23          441
10    14          424
11    20          419
12     0          419
13    10          399
14     5          392
15    16          389
16     9          385
17     1          385
18    13          375
19    19          350
20     2          349
21     4          336
22     6          315
23    15          292

## Most Discounted Payment Methods

In [70]:
dp=td.groupby('Payment Method')['Discount'].sum().sort_values(ascending=False).reset_index()
dp

Payment Method  Discount
0  Mobile carrier Billing  123424.0
1             Credit card  112034.0
2            Prepaid card   47140.0
3              Debit card    9178.0
4           Paypal wallet    2471.0
5          Deirect debits    2429.0
6               Gift card    1777.0

## Product Distribution and its relation with the number of orders

In [73]:
mean=td['Selling_price'].mean()

In [74]:
td['P_Value type']=td['Selling_price'].apply(lambda x:'High Value' if x>=mean else 'Low Value')

In [76]:
td.groupby('P_Value type')['Credit_card'].count()

P_Value type
High Value    5272
Low Value     4722
Name: Credit_card, dtype: int64

## Discount vs. Number of Oders

In [84]:
mc=td.groupby('Merchant_name').count()['Credit_card'].sort_values(ascending=False)
mc

Merchant_name
Barnes & Noble        776
L'Occitane            689
Ulta                  684
Guitar Center         642
The Vitamin Shoppe    622
Cole Hardware         594
DODOcase              578
The North Face        574
REI                   559
Walgreens             552
Road Runner Sports    537
Peninsula Beauty      469
Gymboree              454
Staples               419
Right Start           404
DFLOWERS              377
Nine West             343
PetSmart              336
Costco                231
Nob Hill Foods        154
Name: Credit_card, dtype: int64

In [89]:
md=td.groupby('Merchant_name')['Discount'].sum().sort_values(ascending=False)
md

Merchant_name
Barnes & Noble        23173.0
L'Occitane            21714.0
Ulta                  20607.0
Guitar Center         18886.0
DODOcase              18810.0
The Vitamin Shoppe    17349.0
Cole Hardware         16792.0
The North Face        16601.0
Walgreens             16001.0
REI                   15727.0
Road Runner Sports    15064.0
Peninsula Beauty      14902.0
Gymboree              13986.0
Staples               13061.0
Right Start           12902.0
DFLOWERS              12491.0
PetSmart              10788.0
Nine West              8960.0
Costco                 6313.0
Nob Hill Foods         4326.0
Name: Discount, dtype: float64

### Note : Providing more discounts leads to more number of orders for a merchant